woocommerce/woocommerce
# [ci] Merge test jobs into a single generic job

### Changes proposed in this Pull Request:

Following up on [this comment](https://github.com/woocommerce/woocommerce/pull/45190#issuecomment-1997139478). 
Initially it was just easier to have 4 different types of jobs to move faster without breaking the reporting. Now that we're wrapping up the test workflows consolidation work it makes sense to tackle this and merge all test jobs into a single, generic one.

To make it work, a new `report` object was added to the jobs configuration with the following properties:
- `resultsBlobName`: this should be the name of the artifact to be uploaded (with job index appended to make it unique)
- `resultsPath`: this is the path to the output that will be uploaded under the above name
- `allure`:  a boolean which if true will include the resultsBlobName in the list of reports to be created from the merged artifacts uploaded under the same name.

For every unique `resultsBlobName` that also has the allure property set to true the Publish reports job will run. 

The Allure reports will only be uploaded for the e2e jobs. This should be a temporary measure, until we can update the other end in `woocommerce-test-reports` to not care about the test type.

### How to test the changes in this Pull Request:

Locally, you can test the `ci-jobs` tool by running `pnpm utils ci-jobs`. You can use different options (run --help to see them). 

Expected:
- all test jobs (unit, e2e, api, performance) are returned in the same list.
- there is a new Report jobs list, containing unique entries.

Check CI:
- All types of tests jobs (unit, e2e, api, performance) ran and were successful.
- All jobs that have the report configured uploaded an artifct.
- The 'Archive metrics results' step was only executed for the Metrics job.
- For each unique entry in the Report jobs there is a `Publish reports - <name>`. The job downloaded all artifacts matching the given pattern, successfully generated the Allure report. If the report was for an e2e test suite the 'Publish report to dashboard' was executed. 
- Everything else works as before. Jobs evaluation, Slack alert (tested [here](https://github.com/woocommerce/woocommerce/actions/runs/9398618730/job/25885327540?pr=48175)).
